fix(compiler-cli): incorrectly type checking calls to implicit template variables

crisbeto opened the pull request:
Currently when we encounter an implicit method call (e.g. `{{ foo(1) }}`) and we manage to resolve its receiver to something within the template, we assume that the method is on the receiver itself so we generate a type checking code to reflect it. This assumption is true in most cases, but it breaks down if the call is on an implicit receiver and the receiver itself is being invoked. E.g.

```
<div *ngFor="let fn of functions">{{ fn(1) }}</div>
```

These changes resolve the issue by generating a regular function call if the method call's receiver is pointing to `$implicit`.

Fixes #39634.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1604,15 +1604,21 @@ class TcbExpressionTranslator {
-      addParseSpanInfo(method, ast.nameSpan);
+      const receiver = this.targetToExpression(target, ast.sourceSpan);
+
+      // If the receiver was resolved to `$implicit`, it means that the
+      // implicit variable itself is being invoked so we have to generate
+      // a regular function call, rather than invoking a method on the receiver.
+      const expression = (target.value === '$implicit' && target.name === ast.name) ?
+          wrapForDiagnostics(receiver) :
+          ts.createPropertyAccess(wrapForDiagnostics(receiver), ast.name);
```
NIT: I would just put this block into the `targetToExpression()` helper function. Otherwise, I would rename `targetToExpression()` to `targetToReceiver()`.

petebacondarwin reviewed (approved):
Nice work @crisbeto - it is great when things can be fixed so simply. Thanks for picking up this issue and resolving it so quickly.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1604,15 +1604,21 @@ class TcbExpressionTranslator {
+      // If the receiver was resolved to `$implicit`, it means that the
+      // implicit variable itself is being invoked so we have to generate
+      // a regular function call, rather than invoking a method on the receiver.
+      const expression = (target.value === '$implicit' && target.name === ast.name) ?
+          wrapForDiagnostics(receiver) :
+          ts.createPropertyAccess(wrapForDiagnostics(receiver), ast.name);
```
I think my fix in #33132 was incorrect, as the local template variable name is irrelevant and should not be reflected into the TCB. This is independent from `$implicit`, as e.g. the following should also work:

```ts
export interface Context {
  fn: () => void;
}
```

```html
<ng-template let-function="fn">
  {{ function() }}
</ng-template>
```

should become (in pseudocode)

```ts
var _t1: Context;
var _t2 = _t1.fn;
_t2();
```

The TCB would currently be `_t2.function()` but there is no `function` on the context.

tldr; I believe it's sufficient to only change `const method = wrapForDiagnostics(receiver)` , at which point refactoring the methids is no longer needed.

crisbeto replied:
You're totally right! I'll leave in my test, even though it's basically the same as the one you added in the earlier PR, because the syntax is a little different.

crisbeto committed:
fix(compiler-cli): incorrectly type checking calls to implicit template variables

Currently when we encounter an implicit method call (e.g. `{{ foo(1) }}`) and we manage to resolve
its receiver to something within the template, we assume that the method is on the receiver itself
so we generate a type checking code to reflect it. This assumption is true in most cases, but it
breaks down if the call is on an implicit receiver and the receiver itself is being invoked. E.g.

```
<div *ngFor="let fn of functions">{{ fn(1) }}</div>
```

These changes resolve the issue by generating a regular function call if the method call's receiver
is pointing to `$implicit`.

Fixes #39634.
